From 256e6773b68b460891831aad77b780335ca01cd1 Mon Sep 17 00:00:00 2001
From: Nadav Haklai <nadavh@marvell.com>
Date: Tue, 8 Oct 2013 10:35:09 +0200
Subject: [PATCH 1095/1825] a38x: introduce new machine to the kernel

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit f99cb7d26192c7dffa3a41b1ab1c8936d93bd219

	- update configuration files
	- enable initial support to run on db-a375

Change-Id: I5b2f90f51100f280c9e2b550358c8af781b85522
Signed-off-by: Marcin Wojtas <mw@semihalf.com>
Signed-off-by: Nadav Haklai <nadavh@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/3632
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 arch/arm/Kconfig                |   27 ++++++++++++++++++++++++++-
 arch/arm/Makefile               |    1 +
 arch/arm/boot/compressed/head.S |    3 +++
 arch/arm/kernel/process.c       |    3 +++
 arch/arm/mm/Kconfig             |   10 +++++-----
 arch/arm/plat-armada/Kconfig    |   34 +++++++++++++++++-----------------
 arch/arm/tools/mach-types       |    1 +
 drivers/rtc/Kconfig             |    2 +-
 drivers/tty/serial/8250/Kconfig |    2 +-
 9 files changed, 58 insertions(+), 25 deletions(-)

diff --git a/arch/arm/Kconfig b/arch/arm/Kconfig
index 2bd87c3..3a88b34 100644
--- a/arch/arm/Kconfig
+++ b/arch/arm/Kconfig
@@ -633,6 +633,29 @@ config ARCH_ARMADA370
 	help
 	  Support for the Marvell Armada-370 SoC Family
 
+config ARCH_ARMADA38X
+	bool "Marvell Armada-38x"
+	select PCI
+#	select ARCH_HAS_CPUFREQ
+#	select ARCH_SUPPORTS_MSI
+	select CPU_V7
+	select ARCH_REQUIRE_GPIOLIB
+	select GENERIC_GPIO
+	select GENERIC_TIME
+	select GENERIC_CLOCKEVENTS
+	select PLAT_ARMADA
+	select HAVE_REGS_AND_STACK_ACCESS_API
+	select COMMON_CLKDEV
+	select HAVE_SMP
+	select CLKSRC_MMIO
+	select NEED_MACH_IO_H
+	select NEED_MACH_MEMORY_H
+	select DMABOUNCE
+#	select CACHE_L2X0
+	select MIGHT_HAVE_CACHE_L2X0
+	help
+	  Support for the Marvell Armada-38x SoC Family
+
 config ARCH_ARMADA_XP
 	bool "Marvell Armada XP"
 	select CPU_V7
@@ -1194,6 +1217,7 @@ source "arch/arm/mach-realview/Kconfig"
 source "arch/arm/mach-sa1100/Kconfig"
 
 source "arch/arm/mach-armada370/Kconfig"
+source "arch/arm/mach-armada38x/Kconfig"
 source "arch/arm/mach-armadaxp/Kconfig"
 source "arch/arm/mach-avantalp/Kconfig"
 source "arch/arm/plat-armada/Kconfig"
@@ -1299,7 +1323,8 @@ config XSCALE_PMU
 
 config CPU_HAS_PMU
 	depends on (CPU_V6 || CPU_V6K || CPU_V7 || XSCALE_PMU) && \
-		   (!ARCH_OMAP3 || OMAP3_EMU)
+		   (!ARCH_OMAP3 || OMAP3_EMU) && !ARCH_ARMADA38X
+
 	default y
 	bool
 
diff --git a/arch/arm/Makefile b/arch/arm/Makefile
index 47d2404..c43f223 100644
--- a/arch/arm/Makefile
+++ b/arch/arm/Makefile
@@ -137,6 +137,7 @@ textofs-$(CONFIG_ARCH_MSM8960) := 0x00208000
 # Machine directory name.  This list is sorted alphanumerically
 # by CONFIG_* macro name.
 machine-$(CONFIG_ARCH_ARMADA370)	:= armada370
+machine-$(CONFIG_ARCH_ARMADA38X)	:= armada38x
 machine-$(CONFIG_ARCH_AT91)		:= at91
 machine-$(CONFIG_ARCH_BCMRING)		:= bcmring
 machine-$(CONFIG_ARCH_CLPS711X)		:= clps711x
diff --git a/arch/arm/boot/compressed/head.S b/arch/arm/boot/compressed/head.S
index 392d2069..28063a3 100644
--- a/arch/arm/boot/compressed/head.S
+++ b/arch/arm/boot/compressed/head.S
@@ -12,6 +12,9 @@
 #ifdef CONFIG_ARCH_ARMADA_XP
 #include <mach/armadaxp.h>
 #endif
+#ifdef CONFIG_ARCH_ARMADA38X
+#include <mach/armada38x.h>
+#endif
 
 	.arch	armv7-a
 /*
diff --git a/arch/arm/kernel/process.c b/arch/arm/kernel/process.c
index 843b8e9..0860979 100644
--- a/arch/arm/kernel/process.c
+++ b/arch/arm/kernel/process.c
@@ -115,7 +115,10 @@ static void __soft_restart(void *addr)
 	flush_cache_all();
 
 	/* Turn off caching */
+#if !defined(CONFIG_ARCH_ARMADA38X)
+	/* Temporrary skip the disabling of the caches for Armada 375 until we resolve the reboot hang issue */
 	cpu_proc_fin();
+#endif
 
 	/* Push out any further dirty data, and ensure cache is empty */
 	flush_cache_all();
diff --git a/arch/arm/mm/Kconfig b/arch/arm/mm/Kconfig
index a0e80a9..004245d 100644
--- a/arch/arm/mm/Kconfig
+++ b/arch/arm/mm/Kconfig
@@ -434,7 +434,7 @@ config CPU_V6K
 
 # ARMv7
 config CPU_V7
-	bool "Support ARM V7 processor" if ARCH_INTEGRATOR || MACH_REALVIEW_EB || MACH_REALVIEW_PBX || ARCH_AVANTA_LP
+	bool "Support ARM V7 processor" if ARCH_INTEGRATOR || MACH_REALVIEW_EB || MACH_REALVIEW_PBX || ARCH_AVANTA_LP || ARCH_ARMADA38X
 	select CPU_32v6K
 	select CPU_32v7
 	select CPU_ABRT_EV7
@@ -1201,14 +1201,14 @@ config ENABLE_UNALINGED_ACCESS_FAULT
 
 config AURORA_IO_CACHE_COHERENCY
 	bool "Enable Marvell Aurora I/O cache coherency"
-	depends on ARCH_ARMADA_XP || ARCH_ARMADA370 || ARCH_AVANTA_LP
+	depends on ARCH_ARMADA_XP || ARCH_ARMADA370 || ARCH_AVANTA_LP || ARCH_ARMADA38X
 	default y
 	select HAVE_ARM_SCU
 	help
 	  This option enables the hardware mechanism for I/O cache coherency.
 
 config ALP_IOCC_SYNC_BARRIER_WA
-	depends on ARCH_AVANTA_LP && AURORA_IO_CACHE_COHERENCY
+	depends on (ARCH_AVANTA_LP || ARCH_ARMADA38X) && AURORA_IO_CACHE_COHERENCY
 	bool "Enable IO Sync Barrier WA"
 	default y
 	help
@@ -1256,7 +1256,7 @@ config MV_SUPPORT_L2_DEPOSIT
 
 config MV_SUPPORT_64KB_PAGE_SIZE
 	bool "Support 64KB page size"
-	depends on ARCH_ARMADA_XP || ARCH_ARMADA370
+	depends on ARCH_ARMADA_XP || ARCH_ARMADA370 || ARCH_ARMADA38X
 	default n
 	help
 	  This option enables 64KB pages.
@@ -1279,7 +1279,7 @@ config CACHE_L2X0
 	bool "Enable the L2x0 outer cache controller" if MIGHT_HAVE_CACHE_L2X0
 	default MIGHT_HAVE_CACHE_L2X0
 	select OUTER_CACHE
-	select OUTER_CACHE_SYNC if ARCH_AVANTA_LP && !AURORA_IO_CACHE_COHERENCY
+	select OUTER_CACHE_SYNC if (ARCH_AVANTA_LP || ARCH_ARMADA38X) && !AURORA_IO_CACHE_COHERENCY
 	help
 	  This option enables the L2x0 PrimeCell.
 
diff --git a/arch/arm/plat-armada/Kconfig b/arch/arm/plat-armada/Kconfig
index 97cb75a..16dc8af 100644
--- a/arch/arm/plat-armada/Kconfig
+++ b/arch/arm/plat-armada/Kconfig
@@ -13,10 +13,10 @@ config JTAG_DEBUG
 
 menu "Marvell SoC Included Features"
 
-if ARMADA_XP || AVANTA_LP
+if ARMADA_XP || AVANTA_LP || ARMADA_38X
 config MV_INCLUDE_PEX
 	bool "PCI Express Support"
-	depends on PCI && (MV88F6500 || MV88F6281 || MV78XX0 || ARMADA_XP || AVANTA_LP)
+	depends on PCI && (MV88F6500 || MV88F6281 || MV78XX0 || ARMADA_XP || AVANTA_LP || ARMADA_38X)
 	default y
 	select MV_PEX_2_1X4 if ARMADA_XP
 	select MV_PEX_3_1X4 if ARMADA_XP
@@ -67,22 +67,22 @@ endif
 
 config MV_INCLUDE_PCI
 	bool "PCI Support"
-	depends on PCI && (MV88F6500 || MV88F6281 || MV78XX0 || ARMADA_XP || ARMADA_370 || AVANTA_LP)
+	depends on PCI && (MV88F6500 || MV88F6281 || MV78XX0 || ARMADA_XP || ARMADA_370 || AVANTA_LP || ARMADA_38X)
 	default n
         help
         Please don't change this configs unless you know what you are doing.
 
 config MV_INCLUDE_USB
 	bool "USB Support"
-	depends on MV88F6500 || MV88F6281 || MV78XX0 || ARMADA_XP || ARMADA_370 || AVANTA_LP
+	depends on MV88F6500 || MV88F6281 || MV78XX0 || ARMADA_XP || ARMADA_370 || AVANTA_LP || ARMADA_38X
 	default y
-	select USB_XHCI_PLATFORM if AVANTA_LP
+	select USB_XHCI_PLATFORM if (AVANTA_LP || ARMADA_38X)
         help
         Please don't change this configs unless you know what you are doing.
 
 config MV_INCLUDE_XOR
 	bool "XOR Support"
-	depends on MV88F6500 || MV88F6281 || MV78XX0 || ARMADA_XP || ARMADA_370 || AVANTA_LP
+	depends on MV88F6500 || MV88F6281 || MV78XX0 || ARMADA_XP || ARMADA_370 || AVANTA_LP || ARMADA_38X
 	default y
         help
         Please don't change this configs unless you know what you are doing.
@@ -90,14 +90,14 @@ config MV_INCLUDE_XOR
 config MV_INCLUDE_CESA
 	bool "CESA Support"
 	depends on MV88F6500 || MV88F6082 || MV88F6183 || MV88F6281 || MV78XX0 \
-		|| ARMADA_XP || ARMADA_370 || AVANTA_LP
+		|| ARMADA_XP || ARMADA_370 || AVANTA_LP || ARMADA_38X
 	default y
         help
         Please don't change this configs unless you know what you are doing.
 
 config MV_INCLUDE_NFC
 	bool "Nand Flash Controller Support"
-	depends on MV88F6500 || ARMADA_XP || ARMADA_370 || AVANTA_LP
+	depends on MV88F6500 || ARMADA_XP || ARMADA_370 || AVANTA_LP || ARMADA_38X
 	default y
         help
         Please don't change this configs unless you know what you are doing.
@@ -105,7 +105,7 @@ config MV_INCLUDE_NFC
 config MV_INCLUDE_LEGACY_NAND
 	bool "Legacy NAND Support"
 	depends on MV88F6500 || MV88F6281 || MV78XX0 \
-		|| ARMADA_XP || ARMADA_370 || AVANTA_LP
+		|| ARMADA_XP || ARMADA_370 || AVANTA_LP || ARMADA_38X
 	default y
         help
         Please don't change this configs unless you know what you are doing.
@@ -113,7 +113,7 @@ config MV_INCLUDE_LEGACY_NAND
 config MV_INCLUDE_INTEG_SATA
 	bool "Integrated SATA Support"
 	depends on MV88F6500 || MV88F6281 || MV78XX0 || \
-		   ARMADA_XP || ARMADA_370 || AVANTA_LP
+		   ARMADA_XP || ARMADA_370 || AVANTA_LP || ARMADA_38X
 	default y
         help
         Please don't change this configs unless you know what you are doing.
@@ -121,7 +121,7 @@ config MV_INCLUDE_INTEG_SATA
 config MV_INCLUDE_TDM
 	bool "Integrated TDM Support"
 	depends on MV88F6500 || MV88F6281 || MV78XX0 || \
-		   ARMADA_XP || ARMADA_370 || AVANTA_LP
+		   ARMADA_XP || ARMADA_370 || AVANTA_LP || ARMADA_38X
 	default y
         help
         Please don't change this configs unless you know what you are doing.
@@ -129,13 +129,13 @@ config MV_INCLUDE_TDM
 config MV_INCLUDE_GIG_ETH
 	bool "Giga Ethernet Support"
 	depends on MV88F6500 || MV88F6281 || MV78XX0 || \
-		   ARMADA_XP || ARMADA_370 || AVANTA_LP
+		   ARMADA_XP || ARMADA_370 || AVANTA_LP || ARMADA_38X
 	default y
 
 config MV_INCLUDE_SPI
 	bool "SPI Support"
 	depends on MV88F6500 || MV88F6281 || (MV78XX0 && !MV78XX0_Z0) || \
-		   ARMADA_XP || ARMADA_370 || AVANTA_LP
+		   ARMADA_XP || ARMADA_370 || AVANTA_LP || ARMADA_38X
 	default y
 	help
 	  Please don't change this configs unless you know what you are doing.
@@ -143,14 +143,14 @@ config MV_INCLUDE_SPI
 config MV_INCLUDE_NOR
 	bool "NOR Support"
 	depends on MV88F6500 || MV88F6281 || (MV78XX0 && !MV78XX0_Z0) || \
-		   ARMADA_XP || AVANTA_LP
+		   ARMADA_XP || AVANTA_LP || ARMADA_38X
 	default y
 	help
 	  Please don't change this configs unless you know what you are doing.
 
 config MV_INCLUDE_SDIO
 	bool "SDIO Support"
-	depends on MV88F6500 || MV88F6281 || ARMADA_XP || ARMADA_370 || AVANTA_LP
+	depends on MV88F6500 || MV88F6281 || ARMADA_XP || ARMADA_370 || AVANTA_LP || ARMADA_38X
 	default y
         help
         Please don't change this configs unless you know what you are doing.
@@ -184,7 +184,7 @@ menu "AMP options"
 
 config MV_AMP_ENABLE
         bool "Enable AMP support"
-	depends on (ARMADA_XP || MV78XX0 || AVANTA_LP) && \
+	depends on (ARMADA_XP || MV78XX0 || AVANTA_LP || ARMADA_38X) && \
 		(!ARMADA_XP_REV_Z1) && (SMP) && (!MV_ETH_BM) && \
 		(!ARMADAXP_USE_IRQ_INDIRECT_MODE)
 	default n
@@ -486,7 +486,7 @@ endif
 
 config MV_ETH_PP2
 	tristate "PPv2 driver "
-	depends on ARCH_AVANTA_LP || ARCH_ARMADA375
+	depends on ARCH_AVANTA_LP || ARCH_ARMADA375 || ARMADA_38X
 	default y
 	select MV_ETH_PP2_CLS2
 	select MV_ETH_PP2_CLS3
diff --git a/arch/arm/tools/mach-types b/arch/arm/tools/mach-types
index afbcda2..bb1886f 100644
--- a/arch/arm/tools/mach-types
+++ b/arch/arm/tools/mach-types
@@ -121,6 +121,7 @@ avila			MACH_AVILA		AVILA			526
 xcat98dx		MACH_XCAT98DX		XCAT98DX		527
 armada_xp_fpga		MACH_ARMADA_XP_FPGA	ARMADA_XP_FPGA		528
 avanta_lp		MACH_AVANTA_LP		AVANTA_LP		529
+armada_38x		MACH_ARMADA_38X		ARMADA_38X		530
 feroceon_kw2		ARCH_FEROCEON_KW2	FEROCEON_KW2		529
 edb9302			MACH_EDB9302		EDB9302			538
 husky			MACH_HUSKY		HUSKY			543
diff --git a/drivers/rtc/Kconfig b/drivers/rtc/Kconfig
index 7eda19e..01ac1cf 100644
--- a/drivers/rtc/Kconfig
+++ b/drivers/rtc/Kconfig
@@ -980,7 +980,7 @@ config RTC_DRV_TX4939
 config RTC_DRV_MV
 	tristate "Marvell SoC RTC"
 	depends on ARCH_KIRKWOOD || ARCH_DOVE || \
-		   ARCH_ARMADA_XP || ARCH_ARMADA370 || ARCH_AVANTA_LP
+		   ARCH_ARMADA_XP || ARCH_ARMADA370 || ARCH_AVANTA_LP || ARCH_ARMADA38X
 	help
 	  If you say yes here you will get support for the in-chip RTC
 	  that can be found in some of Marvell's SoC devices, such as
diff --git a/drivers/tty/serial/8250/Kconfig b/drivers/tty/serial/8250/Kconfig
index a24f2d7..706dd9b 100644
--- a/drivers/tty/serial/8250/Kconfig
+++ b/drivers/tty/serial/8250/Kconfig
@@ -274,7 +274,7 @@ config SERIAL_8250_FSL
 
 config SERIAL_8250_DW
 	tristate "Support for Synopsys DesignWare 8250 quirks"
-	depends on SERIAL_8250 && (OF || ARMADA_XP || AVANTA_LP || MV88F6500)
+	depends on SERIAL_8250 && (OF || ARMADA_XP || AVANTA_LP || ARMADA_38X || MV88F6500)
 	help
 	  Selecting this option will enable handling of the extra features
 	  present in the Synopsys DesignWare APB UART.
-- 
1.7.5.4

